server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py

```diff
@@ -456,6 +478,30 @@ def __init__(self, config, weights):
             config, prefix="transformer.wte", weights=weights
         )
 
+        # Buffers need to be persistent to avoid any bug.
+        self.buffers = {}
+        if config.quantize == "gptq-cuda":
+            max_dq_buffer_size = 0
+            for name, submodule in self.named_modules():
+                if isinstance(submodule, (TensorParallelColumnLinear, TensorParallelRowLinear)) and isinstance(submodule.linear, Ex4bitLinear):
+                    max_dq_buffer_size = max(max_dq_buffer_size, submodule.linear.qweight.numel() * 8)
+            
+            intermediate_size = config.n_inner
+            max_seq_len = 2048  # TODO: we should be able to set it
+            
+            self.buffers["temp_state"] = torch.zeros((max_seq_len, intermediate_size), dtype=torch.float16, device=weights.device)
+            self.buffers["temp_dq"] = torch.zeros((1, max_dq_buffer_size), dtype=torch.float16, device=weights.device)
+
+            prepare_buffers(weights.device, self.buffers["temp_state"], self.buffers["temp_dq"])
+
+            # TODO: ability to set them
+            matmul_recons_thd = 8
+            matmul_fused_remap = False
+            matmul_no_half2 = False
+            set_tuning_params(matmul_recons_thd, matmul_fused_remap, matmul_no_half2)
+
+            torch.cuda.empty_cache()
+            
```

Comment:
I moved it to `Model` init. This requires to have access to `model.config` which is currently not defined though. There is `model.transformer.config`, or `model.gpt_neox.config` or `model.model.config` depending on the architecture. Is it intended that the config is not registered at the top level? @OlivierDehaene @Narsil 

The thing is that the `weights = Weights(...)` call is in each model definition, and we need to have loaded all weights to determine the shapes of the buffers. Also, the buffers need to be persistent, while I think this `weights` object is not.

Reply:
The buffers intend to be shared, no ?

So why not just have a single location for these buffers, use the pointer on every layer, and increase the size every time `max_dq_buffer_size = max(max_dq_buffer_size, submodule.linear.qweight.numel() * 8)` is larger ?

The issue with this (and any post loading treatment) is that you're now dealing with updating every single model file, any time of those line hits. This is what we had before and it was painful to maintain.

This seems to be used as globals let's just use them as globals. (They are temporary buffers IIUC preallocated to avoid reallocating them all the time)